kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/CreateCheckpointIterator.java

```diff
@@ -0,0 +1,366 @@
+import static io.delta.kernel.internal.util.Preconditions.checkState;
+
+/**
+ * Replays a history of actions from the transaction log to reconstruct the checkpoint state of the
+ * table. The rules for constructing the checkpoint state are defined in the Delta Protocol:
+ * <a href="https://github.com/delta-io/delta/blob/master/PROTOCOL.md#action-reconciliation">
+ * Checkpoint Reconciliation Rules</a>.
+ * <p>
+ * Currently, the following rules are implemented:
+ * <ul>
+ *     <li> The latest protocol action seen wins </li>
+ *     <li> The latest metaData action seen wins </li>
+ *     <li> For txn actions, the latest version seen for a given appId wins </li>
+ *     <li> Logical files in a table are identified by their (path, deletionVector.uniqueId)
+ *     primary key. File actions (add or remove) reference logical files, and a log can contain
+ *     any number of references to a single file.</li>
+ *     <li>To replay the log, scan all file actions and keep only the newest reference for each
+ *     logical file.</li>
+ *     <li>add actions in the result identify logical files currently present in the table
+ *     (for queries). remove actions in the result identify tombstones of logical files no
+ *     longer present in the table (for VACUUM).</li>
+ *     <li>commit info actions are not included</li>
+ * </ul>
+ * <p>
+ * Following rules are not implemented. They will be implemented as we add support for more
+ * table features over time.
+ * <ul>
+ *     <li> For domainMetadata, the latest domainMetadata seen for a given domain wins.</li>
+ * </ul>
+ */
+public class CreateCheckpointIterator implements CloseableIterator<FilteredColumnarBatch> {
+
+    private static final int[] ADD_ORDINAL = getPathOrdinals(CHECKPOINT_SCHEMA, "add");
+    private static final int[] ADD_PATH_ORDINAL =
+            getPathOrdinals(CHECKPOINT_SCHEMA, "add", "path");
+    private static final int[] ADD_DV_ORDINAL =
+            getPathOrdinals(CHECKPOINT_SCHEMA, "add", "deletionVector");
+
+    private static final int[] REMOVE_ORDINAL = getPathOrdinals(CHECKPOINT_SCHEMA, "remove");
+    private static final int[] REMOVE_PATH_ORDINAL =
+            getPathOrdinals(CHECKPOINT_SCHEMA, "remove", "path");
+    private static final int[] REMOVE_DV_ORDINAL =
+            getPathOrdinals(CHECKPOINT_SCHEMA, "remove", "deletionVector");
+    private static final int[] REMOVE_DELETE_TIMESTAMP_ORDINAL =
+            getPathOrdinals(CHECKPOINT_SCHEMA, "remove", "deletionTimestamp");
+
+    private static final int[] PROTOCOL_ORDINAL = getPathOrdinals(CHECKPOINT_SCHEMA, "protocol");
+    private static final int[] METADATA_ORDINAL = getPathOrdinals(CHECKPOINT_SCHEMA, "metaData");
+    private static final int[] TXN_ORDINAL = getPathOrdinals(CHECKPOINT_SCHEMA, "txn");
+
+    private final TableClient tableClient;
+    private final LogSegment logSegment;
+    private final long minFileRetentionTimestampMillis;
```

Comment:
worth commenting that the tombstone timestamp will be kept if it is strictly greater than this